### PR TITLE
@xmpp

### PR DESCRIPTION
**Pull request by sonnyp**
- use `@xmpp` npm scope
- use eslint (extends standard) with jsx/promise/ava goodies
- simpler code
- es2015
- use [ava](https://github.com/sindresorhus/ava) for tests
- use babel
- use jsx with ltx.createElement pragma
- use es2015 modules
- promises, no moar callbacks
- use [js-sasl](https://github.com/jaredhanson/js-sasl) for SASL
- no more hardcoded/linked features/transports, everything is a "plugin"/module `client.use(plugin)` (see `packages/`)
  - `@xmpp/client` is a preset of common plugins for easy use (~= node-xmpp-client)
  - `@xmpp/client-*` are client related plugins (register, websocket, bind, authentication, stream-management, ...)
- `@xmpp/jid` = `node-xmpp-jid` for now
- `@xmpp/xml` = `ltx` for now
- no more `Iq`, `Message`, `Presence`, plain and row ltx is used
- better control
  - `client.connect()` opens the socket
  - `client.open()` opens the stream
  - `client.bind()` binds the resource
  - FIXME not sure what to name for the helper method that does all previous - we currently override `connect` in `@xmpp/client`
## todo
- tests
- docs
- BOSH

### CODE REVIEW

**Comment by sonnyp** [comment]
# Doubts

Should we make plugins register their dependencies or have the user explicitly use them?

```javascript
client.use(require('@xmpp/client-iq'))
client.use(require('@xmpp/client-session'))

# VS

client.use(require('@xmpp/client-session'))
```

**Comment by sonnyp** [comment]
@hyperlink if you have feedback

I don't know if you use client but I'd like to get the same "architecture" for server/component/... so it might be of your interest anyway

File: packages/client-authentication/index.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+'use strict'
+
+const debug = require('debug')('xmpp:client:authentication')
+
+function authenticate (creds, cb) {
+  const auth = this.authenticators.find(auth => auth.match(this.features))
```
Comment (hyperlink) [inline packages/client-authentication/index.js]: Arrow functions are only supported on node 4 and up. Are we dropping support for 0.12?

Comment (sonnyp) [inline packages/client-authentication/index.js]: we're not dropping support for 0.12 (nor most browsers :) )

It's still to be done but I'll setup babel

**Comment by sonnyp** [comment]
update:

* promise only API
* use ava for tests ( :heart: )
* use eslint (+ react, ava, jsx goodies) instead of standard (but extends standard rules)
* use babel and es2015 modules

**Comment by sonnyp** [comment]
update:

* added component support
* every 'entity' inherits from @xmpp/connection (simplify code and bundle size)

**Comment by astro** [comment]
The `client.authentication()` method doesn't sound like a verb. Can we have `client.authenticate()` or the abbreviated `client.auth()` instead?

**Comment by sonnyp** [comment]
@astro I could't find an "authentication" method.

Are you referring to `@xmpp/client-authentication` ? It's not a method, it's a plugin.

**Comment by sonnyp** [comment]
@astro please feel free to comment on this PR but please be aware this is mostly a proof of concept and I'm splitting it into cleaner PRs

**Comment by sonnyp** [comment]
@astro oh I found that authentication method in the PR description. That was a mistake.

We don't have any authenticate method. We just emit an event like that:

```javascript
client.on('authentication', (authenticate) => {
  authenticate('username', 'password')
})
```

**Pull request closed without merging**